cmd/jujud/updateseries/updateseries.go

```diff
@@ -0,0 +1,396 @@
+		Purpose: "Update Juju agents running on this machine to start with Systemd",
+		Doc:     updateSeriesCommandDoc,
+	}
+}
+
+func (c *UpdateSeriesCommand) Init(args []string) error {
+	switch {
+	case c.toSeries == "" && c.fromSeries == "":
+		return errors.Errorf("both --to-series and --from-series must be specified")
+	case c.toSeries == "":
+		return errors.Errorf("--to-series must be specified")
+	case c.fromSeries == "":
+		return errors.Errorf("--from-series must be specified")
+	case c.toSeries == c.fromSeries:
+		return errors.Errorf("--to-series and --from-series cannot be the same")
+	}
+
+	fromOS, err1 := series.GetOSFromSeries(c.toSeries)
+	toOS, err2 := series.GetOSFromSeries(c.fromSeries)
+	switch {
+	case err1 != nil:
+		return err1
+	case err2 != nil:
+		return err2
+	case fromOS != toOS:
+		return errors.Errorf("series from two different operating systems specified")
+	case fromOS == jujuos.Windows:
+		return errors.NewNotSupported(nil, "windows not supported")
+	}
+
+	ctlr, err := isController()
+	switch {
+	case err != nil:
+		return err
+	case ctlr:
+		return errors.Errorf("cannot run on a controller machine")
+	}
+
+	return c.CommandBase.Init(args)
+}
+
+func (c *UpdateSeriesCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.CommandBase.SetFlags(f)
+	f.StringVar(&c.dataDir, "data-dir", cmdutil.DataDir, "Juju base data directory")
+	f.StringVar(&c.toSeries, "to-series", "", "Series updating to")
+	f.StringVar(&c.fromSeries, "from-series", "", "Series updating from")
+	f.BoolVar(&c.startAgents, "start-agents", false, "start agents if successful with update")
+}
+
+func (c *UpdateSeriesCommand) Run(ctx *cmd.Context) error {
+	if err := c.findAgents(); err != nil {
+		return err
+	}
+
+	fromInitSys, err := service.VersionInitSystem(c.fromSeries)
+	if err != nil {
+		return err
+	}
+
+	toInitSys, err := service.VersionInitSystem(c.toSeries)
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case toInitSys == service.InitSystemUpstart:
+		return errors.NewNotSupported(nil, "downgrade to series using upstart not supported")
+	case toInitSys == fromInitSys && toInitSys == service.InitSystemSystemd:
+		return errors.Errorf("no work to be done, both series use systemd")
+	case toInitSys == service.InitSystemSystemd:
+		errorHappened := false
+		if err = c.writeSystemdAgents(ctx); err != nil {
+			ctx.Warningf("%s", err)
+			errorHappened = true
+		}
+		if err = c.copyTools(); err != nil {
+			ctx.Warningf("%s", err)
+			errorHappened = true
+		} else {
+			ctx.Infof("successfully copied tools and relinked agent tools")
+		}
+		switch {
+		case !errorHappened && c.startAgents:
+			if err = c.startAllAgents(ctx); err != nil {
+				return err
+			}
+			ctx.Infof("all agents successfully restarted")
+		case errorHappened && c.startAgents:
+			return errors.Errorf("anable to start agents due to previous errors")
+		}
+	default:
+		return errors.Errorf("unknown happened")
```

Comment:
Mmm, maybe slightly more descriptive in the message? Technically nothing has happened as we figured out the init systems, but then failed to act on it. Let's at least print the fromInitSys and toInitSys.

return errors.Errorf("Failed to migrate from %s to %s", fromInitSys, toInitSys")

Reply:
done